### PR TITLE
Remove remote prefix from branchname at git checkout if there is only…

### PR DESCRIPTION
**Pull request by DiegoPiloni**
… one remote

This pull request modifies the way in which `checkout_branch` behaves. The new feature removes the remote prefix from remote branches <remote>/<branch> when there is only one remote.

For example:

When `origin` is the only remote, instead of executing `git checkout origin/<branch>` and getting in detached state it executes `git checkout <branch>`.

Issue: #369

### CODE REVIEW

**Comment by nibheis** [comment]
Hello,

I've also been bitten by this 'checkout remote branch gives a detached state' issue. I went through the changes included in this PR, but to be honest I don't get why this fix only applies when there is only one remote.

I tried and tested successfully a different approach: if the branch that the user wants to checkout is remote, then I add the '--track' option while executing the checkout command.

So, I created a "_get_branch_reference(...)" method (using git rev-parse):

```
    def _get_branch_reference(self, branchname, current_path):
        p = Popen(
            ["git", "rev-parse", "--symbolic-full-name", branchname],
            stdout=PIPE,
            stderr=PIPE,
            cwd=os.path.join(self.root_dir, current_path),
        )
        my_output, my_error = p.communicate()
        if p.returncode == 0:
            return my_output.decode("utf-8").strip("\n")
        else:
            return None
```

And then I changed the `checkout_branch(...)` method to look like:

```
    def checkout_branch(self, branchname, current_path):
        """
        Execute git checkout <branch-name> command & return the result.
        Use the --track parameter for a remote branch.
        """
        reference_name = self._get_branch_reference(branchname, current_path)
        if reference_name is None:
            is_remote_branch = False
        else:
            is_remote_branch = self._is_remote_branch(reference_name)

        if is_remote_branch:
            cmd = ["git", "checkout", "--track", branchname]
        else:
            cmd = ["git", "checkout", branchname]

        p = Popen(
            cmd,
            stdout=PIPE,
            stderr=PIPE,
            cwd=os.path.join(self.root_dir, current_path),
        )

        my_output, my_error = p.communicate()
        if p.returncode == 0:
            return { "code": 0, "message": my_output.decode("utf-8") }
        else:
            response = {
                "code": p.returncode,
                "message": my_error.decode("utf-8"),
            }
            response["command"] = " ".join(cmd)
            return response
```

Note that the default behavior of the `checkout_branch()` method is unchanged. In the case of a local branch or in case of error, the current checkout process applies. Also, this works independently of the number of remotes, and does not rely on git guessing correctly what we want to achieve (in the PR, the checkout does the proper branch+tracking only because git detects that there is only one remote). All in all, I think this way of doing it is more robust than the method from the PR.

I cloned jupyterlab-git repo and I added my changes in a branch, meaning that I am ready to send a PR if you think it OK.

**Comment by Leletir** [comment]
Hello, is it possible to have some feedback about this MR ?

**Comment by fcollonval** [comment]
@Leletir, this is missing unit tests. If you are willing to contribute further on this, I can point you to some tests as example.

**Comment by Leletir** [comment]
@fcollonval, for sure, could you explain me which unit test scenario you want to cover ?

**Comment by fcollonval** [comment]
@Leletir thx for proposing your help. Reading more carefully all the comments, I see two actions:

1. Use the better [implementation](https://github.com/jupyterlab/jupyterlab-git/pull/370#issuecomment-514111942) proposed by @nibheis
1. Add unit tests for the two methods `_get_branch_reference` and `checkout_branch`

For the second point, you can take inspiration from 

https://github.com/jupyterlab/jupyterlab-git/blob/98a6cc96f8be48c0ea1ca52127cf1af4d0e77a4f/jupyterlab_git/tests/test_branch.py#L74-L99 

Basically in the code, you need to use explicitly `subprocess.Popen(...)` not `Popen(...)`. Then the *Given* section above shows how to mock the result of `communicate()`. And in the *Then* section, you check the calls and the answer were the expected ones.

Note: you will need to open a new PR.

**Comment by nibheis** [comment]
Hi @fcollonval and @Leletir !

Nice to see some interest in getting this working better.
I had a quick look at how the tests are implemented (I'm new to that).

If I got it correctly:
- the file "tests/test_branch.py" needs to be updated for the new "_get_branch_reference()" method
- a new file (tests/test_checkout.py) needs to be added for the modified "checkout_branch()" method

and probably, some file needs to be changed to call the new test file.

@Leletir, what are your working on, so that we don't duplicate the effort ?

**Comment by nibheis** [comment]
Hi again,

Here are the first tests, for review (tests/test_branch.py). It runs OK ; I just had to change from `Popen` to `subprocess.Popen` in `git.py` for that.

```
@patch('subprocess.Popen')
def test_get_branch_reference_success(mock_subproc_popen):
    actual_response = 0
    branch='test-branch'
    reference = 'refs/remotes/origin/test_branch'
    # Given
    process_mock = Mock()
    attrs = {
        'communicate.return_value': (reference.encode('utf-8'), ''.encode('utf-8')),
        'returncode': 0
    }
    process_mock.configure_mock(**attrs)
    mock_subproc_popen.return_value = process_mock

    # When
    actual_response = Git(root_dir='/bin')._get_branch_reference(
        branchname=branch,
        current_path='test_curr_path')

    # Then
    mock_subproc_popen.assert_has_calls([
        call(
            ['git', 'rev-parse', '--symbolic-full-name', branch],
            stdout=PIPE,
            stderr=PIPE,
            cwd='/bin/test_curr_path'
        ),
        call().communicate()
    ])
    assert actual_response == reference

@patch('subprocess.Popen')
def test_get_branch_reference_faillure(mock_subproc_popen):
    actual_response = 0
    branch='test-branch'
    reference = 'test-branch'
    # Given
    process_mock = Mock()
    attrs = {
        'communicate.return_value': (reference.encode('utf-8'),
        ("fatal: ambiguous argument '"+branch+"': unknown revision or path not in the working tree.").encode('utf-8')),
        'returncode': 128
    }
    process_mock.configure_mock(**attrs)
    mock_subproc_popen.return_value = process_mock

    # When
    actual_response = Git(root_dir='/bin')._get_branch_reference(
        branchname=branch,
        current_path='test_curr_path')

    # Then
    mock_subproc_popen.assert_has_calls([
        call(
            ['git', 'rev-parse', '--symbolic-full-name', branch],
            stdout=PIPE,
            stderr=PIPE,
            cwd='/bin/test_curr_path'
        ),
        call().communicate()
    ])
    assert actual_response is None
```

Please let me know if that is good enough.

**Comment by nibheis** [comment]
I started to have a look at testing `checkout_branch()`... but I have no idea on how to mock the call to `_get_branch_reference()` that is inside. Any clue ?

**Comment by DiegoPiloni** [comment]
+Mariano Garcia Berrotaran <mariano.garciaberrotaran@mercadolibre.com> who
might be interested to collaborate with this PR.

**Comment by nibheis** [comment]
OK... now I am mocking my private method `_get_branch_reference()` and the `subprocess.Popen` call at the same time, just like this:

```
@patch('subprocess.Popen')                                                                                                            
@patch.object(Git, '_get_branch_reference', return_value=None)
def test_checkout_branch_without_ref_success(mock__get_branch_reference, mock_subproc_popen):                                         
    branch='test-branch'                                                                                                              
            
    # Given 
    process_mock = Mock()                                                                                                             
    attrs = {
        'communicate.return_value': ('checkout output from git'.encode('utf-8'), ''.encode('utf-8')),                                 
        'returncode': 0
    }   
    process_mock.configure_mock(**attrs)
    mock_subproc_popen.return_value = process_mock                                                                                    
            
    # When      
    actual_response = Git(root_dir='/bin').checkout_branch(
        branchname=branch, 
        current_path='test_curr_path')                                                                                                
                                                                                                                                      
    # Then
    mock_subproc_popen.assert_has_calls([                                                                                             
        call(
            ['git', 'checkout', branch],                                                                                              
            stdout=PIPE,                                                                                                              
            stderr=PIPE,
            cwd='/bin/test_curr_path'                                                                                                 
        ),
        call().communicate()
    ])      
    assert { "code": 0, "message": "checkout output from git" } == actual_response                                                    
            
                                                                                                                                      
@patch('subprocess.Popen')
@patch.object(Git, '_get_branch_reference', return_value=None)                                                                        
def test_checkout_branch_without_ref_failure(mock__get_branch_reference, mock_subproc_popen):                                         
    branch='test-branch'

    # Given
    process_mock = Mock() 
    attrs = {
        'communicate.return_value': (''.encode('utf-8'),                                                                              
            ("error: pathspec '"+branch+"' did not match any file(s) known to git").encode('utf-8')),                                 
        'returncode': 1                                                                                                               
    }   
    process_mock.configure_mock(**attrs)
    mock_subproc_popen.return_value = process_mock                                                                                    
            
    # When                                                                                                                            
    actual_response = Git(root_dir='/bin').checkout_branch(                                                                           
        branchname=branch,                                                                                                            
        current_path='test_curr_path')                                                                                                
            
    # Then  
    mock_subproc_popen.assert_has_calls([                                                                                             
        call(                                                                                                                         
            ['git', 'checkout', branch],
            stdout=PIPE,
            stderr=PIPE,
            cwd='/bin/test_curr_path'                                                                                                 
        ),  
        call().communicate()
    ])          
    assert {    "code": 1,                                                                                                            
                "message": "error: pathspec '"+branch+"' did not match any file(s) known to git",                                     
                "command": "git checkout "+branch} == actual_response
```

Works as expected. We need at least 4 more functions to test all possibilities.
Is that the right way of doing it ?

Where should I place those tests for the method `checkout_branch()`? In the same test file (`tests/test_branch.py`) - I think it's OK.

**Comment by fcollonval** [comment]
@nibheis 
Don't hesitate to create a new PR with your code. And write here you are working on this in the PR ###.

Code in comments are not very useful as github tooling is not handling them: no review, no automatic code testing, no in-code comment... So shoot that promising code in a PR labelled as WIP (work in progress) if you have doubts 😉 

> Is that the right way of doing it ?

Yes

> Where should I place those tests for the method checkout_branch()? In the same test file (tests/test_branch.py) - I think it's OK.

Yes

**Comment by Leletir** [comment]
@nibheis which additional scenarios do you want to test ?
I could help you to get this done.

**Comment by nibheis** [comment]
@Leletir, I'm done adding all the tests:

```
jupyterlab_git/tests/test_branch.py::test_checkout_branch_noref_success PASSED
jupyterlab_git/tests/test_branch.py::test_checkout_branch_noref_failure PASSED
jupyterlab_git/tests/test_branch.py::test_checkout_branch_remoteref_success PASSED
jupyterlab_git/tests/test_branch.py::test_checkout_branch_headsref_failure PASSED
jupyterlab_git/tests/test_branch.py::test_checkout_branch_headsref_success PASSED
jupyterlab_git/tests/test_branch.py::test_checkout_branch_remoteref_failure PASSED
jupyterlab_git/tests/test_branch.py::test_get_branch_reference_success PASSED
jupyterlab_git/tests/test_branch.py::test_get_branch_reference_faillure PASSED
```

@fcollonval, I rebased my branch to the latest master, everything is commit+push in it.
PR385 (https://github.com/jupyterlab/jupyterlab-git/pull/385) has been updated accordingly.

**Comment by Leletir** [comment]
Who can review your PR ?

**Comment by fcollonval** [comment]
Superseeded by #385 => closing this
